[clang][clang-scan-deps] Add named modules to format 'experimental-full'

naveen-seth opened the pull request:
This adds standard C++ named modules to the scanning output format 'experimental-full'.

This is motivated by an effort to provide native support for explicit module builds from the Clang driver.
The driver support for explicit module builds is planned to support combined usage of Clang modules and C++ named modules. With this, only a single scan is required to capture dependencies for both module types.

naveen-seth committed:
[clang][clang-scan-deps] Add named modules to format 'experimental-full'

This adds standard C++ named modules to the scanning output format
'experimental-full'.

This is motivated by an effort to provide native support for explicit
module builds from the Clang driver.
The driver support for explicit module builds is planned to support
combined usage of Clang modules and C++ named modules.
With this, only a single scan is required to capture dependencies for
both module types.

ChuanqiXu9 reviewed (commented):
I don't feel bad but I'd like to leave the formal approvals to @jansvoboda11 and @Bigcheese here.

naveen-seth committed:
Omit 'named-module-deps: […]' when empty to avoid updating all tests

naveen-seth committed:
Avoid implicit conversion in non-empty check

naveen-seth committed:
"module-name" -> "named-module"

The corresponding variable name is also updated as it was also
introduced in this PR.

naveen-seth committed:
Let consumers decide whether to handle named module scan results

jansvoboda11 commented on clang/test/ClangScanDeps/modules-full-named-modules.cppm:
```diff
@@ -0,0 +1,347 @@
+// This checks that clang-scan-deps properly outputs named module dependencies 
+// when using the the scanning output format 'experimental-full'.
+//
+// See commit 72304.
```
```suggestion
// See PR #72304.
```

jansvoboda11 commented on clang/test/ClangScanDeps/modules-full-named-modules.cppm:
```diff
@@ -0,0 +1,347 @@
+// See commit 72304.
+// UNSUPPORTED: target={{.*}}-aix{{.*}}
+//
+// The slash direction in linux and windows are different.
+// UNSUPPORTED: system-windows
```
We typically deal with separator differences by normalizing them to `/` like so:
```shell
// RUN: cat %t/result.json | sed 's:\\\\\?:/:g' | FileCheck -DPREFIX=%/t %s
```
Would something like that work for this test?

I see most of your matches are fairly simple (e.g. `[[PREFIX]]/M.cppm`) and there are no complex search paths involved, so maybe matching with `{{.*}}M.cppm` would do the trick too.

naveen-seth committed:
Fix wording mistake in comment

naveen-seth committed:
Make test windows-compatible

jansvoboda11 reviewed (approved):
LGTM, but let's wait for @Bigcheese to take a look as well.

shafik commented on clang/include/clang/Tooling/DependencyScanning/DependencyScanningTool.h:
```diff
@@ -188,13 +192,23 @@ class FullDependencyConsumer : public DependencyConsumer {
+  void handleProvidedAndRequiredStdCXXModules(
+      std::optional<P1689ModuleInfo> Provided,
+      std::vector<P1689ModuleInfo> Requires) override {
+    ModuleName = Provided ? Provided->ModuleName : "";
```
Static analysis flagged code around here and I am stuck trying to understand what `ModuleName` is being used for here. Is this dead code? If not there should really be a comment explaining how it is used.

naveen-seth replied:
`ModuleName` stores the name of the exported standard C++ module.
After `takeTranslationUnitDeps()`, it is stored in the `TranslationUnitDeps.ID.ModuleName` field [(here)](
https://github.com/naveen-seth/llvm-project/blob/74772e0e1b54f8f2648c1bcc0311d3389156ab7c/clang/lib/Tooling/DependencyScanning/DependencyScanningTool.cpp#L174).

I don't think this should be dead code because it is used in `ClangScanDeps.cpp` when printing the `--experimental-full` format (first reassigned [here](https://github.com/naveen-seth/llvm-project/blob/4936fc5a5663717ef6f3617320875b07fa201304/clang/tools/clang-scan-deps/ClangScanDeps.cpp#L407C8-L407C19), then used for printing [here](https://github.com/naveen-seth/llvm-project/blob/4936fc5a5663717ef6f3617320875b07fa201304/clang/tools/clang-scan-deps/ClangScanDeps.cpp#L525-L526)).

Maybe it would improve clarity if we keep `ContextHash` and `ModuleName` together in a `ModuleID` field within `FullDependencyConsumer`, just as TranslationUnitDeps does, rather than keeping them ungrouped.
This would then directly lead people to the documentation of `ModuleID.ModuleName` [(here)](https://github.com/naveen-seth/llvm-project/blob/4936fc5a5663717ef6f3617320875b07fa201304/clang/include/clang/Tooling/DependencyScanning/ModuleDepCollector.h#L93).

shafik reviewed (commented):
Apologies for the after commit review.
